use dev.dirs to calculate terasology directories

soloturn opened the pull request:
the library works according to the XDG standard, and offers translations for linux, macosx, windows. see for details:

https://github.com/dirs-dev/directories-jvm


so ... 
```
❯ ls -la /home/st/.local/share/terasology/
gradle game
...
❯ ls -la /home/st/.local/share/terasology/
.rw-r--r-- 9.3k rt 12 Okt 17:56 config.cfg
drwxr-xr-x    - rt 12 Okt 17:56 configs
drwxr-xr-x    - rt 12 Okt 17:56 logs
drwxr-xr-x    - rt 12 Okt 17:56 modules
drwxr-xr-x    - rt 12 Okt 17:56 recordings
drwxr-xr-x    - rt 12 Okt 17:56 sandbox
drwxr-xr-x    - rt 12 Okt 17:56 saves
drwxr-xr-x    - rt 12 Okt 17:56 screenshots
drwxr-xr-x    - rt 12 Okt 17:56 shaders

❯ export XDG_DATA_HOME /tmp/xdg-for-test 
❯ ls -la /tmp/xdg-for-test/terasology/
.rw-r--r-- 9.3k rt 12 Okt 18:04 config.cfg
drwxr-xr-x    - rt 12 Okt 18:04 configs
drwxr-xr-x    - rt 12 Okt 18:04 logs
drwxr-xr-x    - rt 12 Okt 18:04 modules
drwxr-xr-x    - rt 12 Okt 18:04 recordings
drwxr-xr-x    - rt 12 Okt 18:04 sandbox
drwxr-xr-x    - rt 12 Okt 18:04 saves
drwxr-xr-x    - rt 12 Okt 18:04 screenshots
drwxr-xr-x    - rt 12 Okt 18:04 shaders

```

soloturn committed:
use dev.dirs to calculate terasology directories

the library works according to the XDG standard, and offers translations
for linux, macosx, windows. see for details:

https://github.com/dirs-dev/directories-jvm

fixes #5281.

Co-authored-by: BenjaminAmos <24301287+BenjaminAmos@users.noreply.github.com>

BenjaminAmos commented:
It looks like the `directories-jvm` library uses a Powershell script to try and identify special folders on Windows, which may fail in restricted environments (see also the pinned issue `https://github.com/dirs-dev/directories-jvm/issues/49`). In general the library would probably work well for macOS and Linux but I have reservations about it continuing to work under Windows. Our current approach using JNA to call the supported Win32 API is much more robust.

This may still be the better approach overall but I need to be sure that it will not make things worse for Windows users.

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -33,15 +32,14 @@
-    private static final String TERASOLOGY_FOLDER_NAME = "Terasology";
-    private static final Path LINUX_HOME_SUBPATH = Paths.get(".local", "share", "terasology");
-
+    private static final ProjectDirectories projDirs = ProjectDirectories.from("org", "terasology", "terasology");
```
```suggestion
    private static final ProjectDirectories PROJECT_DIRS = ProjectDirectories.from("org", "terasology", "terasology");
```

Checkstyle is complaining about the naming style here. I also think "proj" is a bit vague for an abbreviated form of "project".

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -33,15 +32,14 @@
-    private static final Path LINUX_HOME_SUBPATH = Paths.get(".local", "share", "terasology");
-
+    private static final ProjectDirectories projDirs = ProjectDirectories.from("org", "terasology", "terasology");
+    private static final Path PROJECT_PATH = Paths.get(projDirs.dataDir);
```
```suggestion
    private static final Path PROJECT_PATH = Paths.get(PROJECT_DIRS.dataDir);
```

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -33,15 +32,14 @@
-    private static final String LOG_DIR = "logs";
-    private static final String SHADER_LOG_DIR = "shaders";
+    private static final String LOG_DIR = projDirs.dataLocalDir + "/logs";
+    private static final String SHADER_LOG_DIR = projDirs.dataLocalDir + "/shaders";
```
```suggestion
    private static final String LOG_DIR = PROJECT_DIRS.dataLocalDir + "/logs";
    private static final String SHADER_LOG_DIR = PROJECT_DIRS.dataLocalDir + "/shaders";
```

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -33,15 +32,14 @@
+    private static final String SHADER_LOG_DIR = projDirs.dataLocalDir + "/shaders";
     private static final String MODULE_DIR = "modules";
-    private static final String MODULE_CACHE_DIR = "cachedModules";
+    private static final String MODULE_CACHE_DIR = projDirs.cacheDir + "/cachedModules";
```
```suggestion
    private static final String MODULE_CACHE_DIR = PROJECT_DIRS.cacheDir + "/cachedModules";
```

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -67,7 +65,7 @@ public final class PathManager {
     private PathManager() {
         installPath = findInstallPath();
-        homePath = installPath;
+        homePath = PROJECT_PATH;
```
```suggestion
        homePath = installPath;
```
`findInstallPath()` provides a fallback path for unusual scenarios that I think we should keep. Defaulting to the current directory is normal for development anyway, I think. You can default to `PROJECT_PATH` in `findInstallPath` if needed.

soloturn replied:
returning projectpath in a function called findinstallpath i thought is too much :) but let me digest your suggestion ...

soloturn replied:
Already matches this - the constructor now sets `homePath = installPath;` as the fallback default (kept in step with `findInstallPath()`'s own current-directory fallback), and `findInstallPath()` itself is untouched. Resolving.

BenjaminAmos commented on facades/PC/src/main/java/org/terasology/engine/Terasology.java:
```diff
@@ -274,14 +274,7 @@ private void handleLaunchArguments() throws IOException {
-        } else {
-            PathManager.getInstance().useDefaultHomePath();
-        }
+        PathManager.getInstance().useDefaultHomePath();
```
This should not have been removed. Having the ability to override the game's home directory is important for testing with multiple clients.

soloturn replied:
how is this test done, or in other words, why setting XDG environment variables is not good enough?

soloturn replied:
`--homedir` is still here (`Terasology.java` line 131-132, wired to `useOverrideHomePath` at line 279) - whatever revision this was flagged against must have dropped it in an intermediate diff. Current branch keeps it. Resolving.

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -316,8 +293,8 @@ public Path getSandboxPath() {
-        logPath = homePath.resolve(LOG_DIR);
-        shaderLogPath = logPath.resolve(SHADER_LOG_DIR);
+        logPath = Paths.get(LOG_DIR);
+        shaderLogPath = Paths.get(SHADER_LOG_DIR);
```
Why are `logPath` and `shaderLogPath` absolute paths when all the other paths are relative to the home directory?

soloturn replied:
> Why are `logPath` and `shaderLogPath` absolute paths when all the other paths are relative to the home directory?

the reason why i like the new appdir / xdg / macos directory structure so much is that one can properly separate stuff which should be in a backup and what not. config i want to backup. shared data as well. local data not, but depends on the use case. cache and logs i never want a backup. coming back to your question: i did not see any harm at the moment to leave e.g. config in homedirectory, even it is not xdg compliant.

soloturn replied:
Also already addressed - `LOG_DIR`/`SHADER_LOG_DIR` are just relative directory names now (`"logs"`/`"shaders"`), and `logPath`/`shaderLogPath` are built the same way as every other path here: `logBase.resolve(LOG_DIR)` off an OS-standard-or-homePath base (see `updateDirs()`). Nothing hardcodes an absolute string against `PROJECT_DIRS` directly anymore. Resolving.

BenjaminAmos reviewed (changes requested):
At a cursory glance, these changes appear to break several existing workflows, especially those using a different home directory.

soloturn commented:
> It looks like the `directories-jvm` library uses a Powershell script to try and identify special folders on Windows, which may fail in restricted environments (see also the pinned issue `https://github.com/dirs-dev/directories-jvm/issues/49`). In general the library would probably work well for macOS and Linux but I have reservations about it continuing to work under Windows. Our current approach using JNA to call the supported Win32 API is much more robust.
> 
> This may still be the better approach overall but I need to be sure that it will not make things worse for Windows users.

i share your concern, and put a comment on https://github.com/dirs-dev/directories-jvm/pull/61,  am thinking that one could make a pull request towards that library even ... and, i left a rant at microsofts openjdk, which i am less fond of: https://github.com/microsoft/openjdk/issues/628.

sideeffffect commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -316,8 +293,8 @@ public Path getSandboxPath() {
+        shaderLogPath = Paths.get(SHADER_LOG_DIR);
         screenshotPath = homePath.resolve(SCREENSHOT_DIR);
         nativesPath = installPath.resolve(NATIVES_DIR);
         configsPath = homePath.resolve(CONFIGS_DIR);
```
```suggestion
        configsPath = Paths.get(CONFIGS_DIR);
```

soloturn replied:
Same reasoning as the sibling comment on `CONFIGS_DIR` below - `configsPath` is deliberately computed from a conditional base (`usingOsStandardDirs ? PROJECT_DIRS.configDir : homePath`) so it keeps nesting under an overridden home path instead of always defaulting to the OS-standard location. No change.

sideeffffect commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -33,15 +32,14 @@
+    private static final String MODULE_CACHE_DIR = PROJECT_DIRS.cacheDir + "/cachedModules";
     private static final String SCREENSHOT_DIR = "screenshots";
     private static final String NATIVES_DIR = "natives";
     private static final String CONFIGS_DIR = "configs";
```
```suggestion
    private static final String CONFIGS_DIR = PROJECT_DIRS.configDir;
```

soloturn replied:
No change - this is now deliberately more nuanced than a direct `PROJECT_DIRS.configDir` assignment. `updateDirs()` picks the config base conditionally: `usingOsStandardDirs ? Paths.get(PROJECT_DIRS.configDir) : homePath`. Always using `PROJECT_DIRS.configDir` (as this suggestion would) breaks that - under `--homedir`/a manually-picked home, config would keep landing in the OS-standard location instead of nesting under the chosen home like everything else, which is exactly the regression `overrideHomePathIsRespectedByLogAndModuleCachePaths` guards against for logs and the module cache. Leaving `CONFIGS_DIR` as a plain relative name and resolving it against the conditional base is what makes that possible.

soloturn commented:
Pushed a fix for a real regression this branch introduced: `--homedir` (TerasologyLauncher's own mechanism for pointing the game at a custom home directory - portable installs, testing multiple clients against separate home directories) stopped controlling where **logs, shader logs, and the module cache** end up.

`LOG_DIR`/`SHADER_LOG_DIR`/`MODULE_CACHE_DIR` were computed as absolute OS-standard paths (`dataLocalDir`/`cacheDir`) rather than plain relative names, so `updateDirs()`'s `Paths.get(LOG_DIR)` and `homePath.resolve(MODULE_CACHE_DIR)` (`resolve()` on an already-absolute path just returns it unchanged) both ignored `homePath` entirely - regardless of whether it came from `useDefaultHomePath()` or `useOverrideHomePath()`. Confirmed by tracing `TerasologyLauncher`'s `GameStarter`: it always launches with `--homedir=<its own chosen path>`, so this affected every launcher-managed install, not just direct/manual runs.

Reverted those three to plain relative names, resolved against `homePath` in `updateDirs()` exactly as before this branch (shader logs nested under the log dir, as originally). `useDefaultHomePath()` itself - the actual point of this PR - is untouched.

Added `overrideHomePathIsRespectedByLogAndModuleCachePaths()` to `PathManagerTest` to catch this specifically. `PathManagerTest`: 9/9 in `unitTest`, 1/1 in `filesystemSideEffectTest`. Also spot-checked on disk that the filesystem-side-effect test now creates `logs/`, `logs/shaders/`, and `cachedModules/` directly under the dev.dirs home directory rather than scattered across `dataLocalDir`/`cacheDir`.

soloturn commented:
Simple version of above, for anyone skimming:

Launcher always say "put stuff here" (`--homedir`). Game listen. All folders - saves, logs, shaders, module cache, screenshots, configs - go under that one folder launcher pick. Same as before this PR. My fix above make sure of that (before fix, logs and module cache secretly ignored `--homedir` and went wrong place anyway).

Tests work same way: `PathManagerTest` set directory manually too (`useOverrideHomePath`), same as launcher does. Manual-set path never changed behavior here - just fixed a bug where 3 sub-folders ignored it.

Only thing that DID change: what folder game pick for itself when NOBODY tells it where to go (`useDefaultHomePath`, no `--homedir` flag). Old default: Windows = Saved Games folder, Mac = Library/Application Support/Terasology. New default: OS standard app folder via dev.dirs, different name/place. That's the ONE test tagged `filesystem-side-effects` covers.

Launcher never uses that default path - launcher always picks folder itself. So: launcher users, nothing change. People running game directly with no launcher and no `--homedir` flag, default folder changes.

soloturn commented:
Simpler:

Launcher tells game: "save here". Game saves everything there. Always did. Still does. No change for launcher.

Only change: if nobody tells game where to save (no launcher, no flag), game picks a different default folder now.

soloturn commented:
More detail, simple words:

Launcher starts game with flag `--homedir=<folder>`. Game puts saves, logs, shaders, module cache, everything in that folder. Always did this. Still does. No change for launcher.

Test for this: `PathManagerTest`, method `overrideHomePathIsRespectedByLogAndModuleCachePaths`. It sets a folder manually (same way launcher does with `--homedir`), then checks logs and module cache land inside that folder. This test did not exist before - I added it in my fix commit, because the old test only checked saves/screenshots/configs land in the right folder, not logs/module cache. That gap is why the bug (logs/module cache ignoring `--homedir`) was not caught earlier.

Only change: if game starts with NO `--homedir` flag (direct run, no launcher), it picks a different default folder now. Test for that: `useDefaultHomePathResolvesToARealProjectDirectory`, tagged `filesystem-side-effects` (existed before, still passes).

soloturn commented:
Launcher and tests: set folder manually (`--homedir` / `useOverrideHomePath`). No change from this PR.

No `--homedir` flag passed (direct run, no launcher) → OS standard folder (dev.dirs):

| OS | Folder |
|---|---|
| Linux | `~/.local/share/terasology` |
| macOS | `~/Library/Application Support/org.terasology.terasology` |
| Windows | `%APPDATA%\org\terasology\terasology\data` |
| Launcher (any OS) | wherever launcher picked, via `--homedir` - not OS standard, launcher's own choice |

soloturn commented:
added table to request text

soloturn commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -33,15 +32,14 @@
+    private static final String SHADER_LOG_DIR = PROJECT_DIRS.dataLocalDir + "/shaders";
     private static final String MODULE_DIR = "modules";
-    private static final String MODULE_CACHE_DIR = "cachedModules";
+    private static final String MODULE_CACHE_DIR = PROJECT_DIRS.cacheDir + "/cachedModules";
```
No change for now. This is real as far as it goes - `dev.dirs` shells out to PowerShell on Windows and a missing/blocked install could plausibly surface as a null or malformed field - but I can't confirm it actually happens in practice (PowerShell has shipped with Windows since Vista; this would need a locked-down/stripped environment), and a proper fix means deciding what `PathManager` should *do* about it: fail fast with a clear error at startup, fall back to a hardcoded default, or something else. That's a design call beyond a targeted fix, and I'd rather not guess at it without being able to reproduce the failure or exercise the fallback path. Flagging it as a known follow-up rather than fixing blind. Closing this thread.

soloturn commented on engine-tests/src/test/java/org/terasology/engine/core/PathManagerTest.java:
```diff
@@ -70,6 +68,25 @@ public void overrideHomePathCreatesDirectories() {
+        assertTrue(pathManager.getShaderLogPath().startsWith(homePath),
+                "Expected shader log path to nest under the overridden home path: " + pathManager.getShaderLogPath());
+        assertTrue(pathManager.getModulePaths().stream().anyMatch(path -> path.startsWith(homePath)),
+                "Expected the module cache path to nest under the overridden home path, among: " + pathManager.getModulePaths());
```
Fixed as suggested - `getModulePaths()` does contain both the home module path and the module-cache path under override, so `anyMatch(startsWith(homePath))` could pass on the former alone even if the cache path regressed. Now asserts `homePath.resolve("cachedModules")` is actually present in commit 8e7dacd9f.

soloturn commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -400,7 +407,10 @@ private static String nativeArchName() {
     protected ImmutableList<Path> defaultModPaths() throws IOException {
         Path homeModPath = homePath.resolve(MODULE_DIR);
-        Path modCachePath = homePath.resolve(MODULE_CACHE_DIR);
+        // Same OS-standard-vs-homePath split as logPath in updateDirs(): the module cache is a
+        // cache (cacheDir, e.g. ~/.cache on Linux) only while homePath is still the OS default.
+        Path modCacheBase = usingOsStandardDirs ? Paths.get(PROJECT_DIRS.cacheDir) : homePath;
+        Path modCachePath = modCacheBase.resolve(MODULE_CACHE_DIR);
```
Confirmed real, and worse than "can" - it always returned the wrong thing: `modPaths.get(0)` is the module-cache path (`homePath == installPath`) or the install path (otherwise), never `homeModPath`. Every actual caller (`ModuleInstaller`, `ClientConnectionHandler`, `Behavior[/Collective]System`) wants the user's home module directory specifically, to install/save into - not the read-only install dir or the OS cache. Fixed to return `homePath.resolve(MODULE_DIR)` directly in commit 8e7dacd9f, with a regression test (`getHomeModPathReturnsHomeDirectorysOwnModulesFolder`).

soloturn commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -398,9 +407,31 @@ private static String nativeArchName() {
+        if (OS.get() != OS.LINUX) {
+            return Paths.get(PROJECT_DIRS.dataLocalDir);
+        }
+        String xdgStateHome = System.getenv("XDG_STATE_HOME");
+        Path stateHome = (xdgStateHome != null && !xdgStateHome.isEmpty())
+                ? Paths.get(xdgStateHome)
+                : Paths.get(System.getProperty("user.home"), ".local", "state");
```
Fixed in commit 8e7dacd9f - `resolveStateDir()` now also checks `Paths.get(xdgStateHome).isAbsolute()` before using it, falling back to `~/.local/state` otherwise, per spec. Didn't add a dedicated unit test for the relative-value branch specifically: `System.getenv()` isn't settable in-process without reflection hacks, and reworking `resolveStateDir()` to take an injectable environment lookup just for this felt like more surface area than the fix warranted. The existing `PathManagerTest` suite (which exercises `resolveStateDir()` indirectly through `getLogPath()`) still passes.
